### PR TITLE
[JENKINS-67602] Skip corrupted fingerprint files

### PR DESCRIPTION
**Pull request by blackliner**
See [JENKINS-67602](https://issues.jenkins-ci.org/browse/JENKINS-67602).

### Issue

We got a NPE because apparently the fingerprint xml files were corrupted. We had to delete them (`$JENKINS_HOME/fingerprints`) for Jenkins to work again. Maybe this could also be expanded by a nice error message, or actually checking the xml's at some other place. I have no experience in this codebase.

### Proposed changelog entries

* Skip corrupted fingerprint files

### log

```
14:34:12  java.lang.NullPointerException
14:34:12  	at hudson.model.Fingerprint.initFacets(Fingerprint.java:1412)
14:34:12  	at hudson.model.Fingerprint.load(Fingerprint.java:1355)
14:34:12  	at hudson.model.FingerprintMap.load(FingerprintMap.java:91)
14:34:12  	at hudson.model.FingerprintMap.load(FingerprintMap.java:46)
14:34:12  	at hudson.util.KeyedDataStorage.get(KeyedDataStorage.java:160)
14:34:12  	at hudson.model.FingerprintMap.get(FingerprintMap.java:81)
14:34:12  	at hudson.model.FingerprintMap.get(FingerprintMap.java:46)
14:34:12  	at hudson.util.KeyedDataStorage.getOrCreate(KeyedDataStorage.java:110)
14:34:12  	at hudson.model.FingerprintMap.getOrCreate(FingerprintMap.java:67)
14:34:12  	at hudson.model.FingerprintMap.getOrCreate(FingerprintMap.java:63)
14:34:12  	at com.cloudbees.plugins.credentials.CredentialsProvider.getOrCreateFingerprintOf(CredentialsProvider.java:1403)
14:34:12  	at com.cloudbees.plugins.credentials.CredentialsProvider.trackAll(CredentialsProvider.java:1459)
14:34:12  	at com.cloudbees.plugins.credentials.CredentialsProvider.track(CredentialsProvider.java:1421)
14:34:12  	at org.jenkinsci.plugins.credentialsbinding.MultiBinding.getCredentials(MultiBinding.java:200)
14:34:12  	at org.jenkinsci.plugins.credentialsbinding.impl.StringBinding.bindSingle(StringBinding.java:62)
14:34:12  	at org.jenkinsci.plugins.credentialsbinding.Binding.bind(Binding.java:150)
14:34:12  	at org.jenkinsci.plugins.credentialsbinding.impl.BindingStep$Execution2.doStart(BindingStep.java:134)
14:34:12  	at org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecution.lambda$run$0(GeneralNonBlockingStepExecution.java:77)
14:34:12  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
14:34:12  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
14:34:12  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
14:34:12  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
14:34:12  	at java.base/java.lang.Thread.run(Thread.java:829)
14:34:12  Finished: FAILURE
```

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by basil — COMMENTED** [review]
Seems to be the same issue reported in [JENKINS-67602](https://issues.jenkins.io/browse/JENKINS-67602). Ideally, we would update that bug with steps to reproduce the issue from scratch and then detrmine the root cause of the problem, which is that the fingerprints are getting (erroneously!) corrupted in the first place. In the absence of a root cause, I think it is fine to skip over fingerprints with null facets, but we should at least log a warning mentioning JENKINS-67602. If we log a warning, we have a chance of noticing the problem in the future and possibly finding the root cause. If we drive on without logging anything, we have no chance of ever finding the root cause.

**Comment by blackliner** [comment]
Sorry, I am not familiar with the codebase, how you do logging, or with java to be quite frank (last time is about 15y ago 😓 )

Don't get me wrong, logging would be the right way, I agree, but I have no clue how to do it.

File: core/src/main/java/hudson/model/Fingerprint.java
Comment (timja) [inline core/src/main/java/hudson/model/Fingerprint.java:1417]: something like:

```suggestion
        if (fingerprint == null) {
            return;
        }
        if (fingerprint.facets == null) {
            logger.log(Level.WARNING, "Missing fingerprint facet, see JENKINS-67602");
            return;
        }

```

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by basil — APPROVED** [review]

**Pull request merged**
